# **Algorithmic Methods for Data Mining: Homework 3**

**Author:** Miguel Angel Sanchez Cortes

*MSc. in Data Science, Sapienza University of Rome*

---

## **0. Uploading the Classes and Modules**

Before doing any kind of analysis it is necessary to upload both the relevant Classes and Modules we will use to work.

In [1]:
from modules.search_engine import SearchEngine, TopKSearchEngine, WeightedTopKSearchEngine
from modules.data_preprocesser import DataPreprocesser
from modules.web_scraper import WebScraper
from modules.html_parser import HTMLParser
from modules.map_plotter import MapPlotter
import pickle



---

## **1. Data Collection**

For this homework, there was no provided dataset, instead, we had to build our own. We were asked to obtain relevant information for the courses contained in the [Find a Masters](https://www.findamasters.com/) website. Specifically, we were asked to obtain information for the Masters courses contained on the first **400** pages of the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website.

To do this, we created two custom-made **Python classes** that performed *Web Scraping* and *HTML Parsing* on the first 400 pages of the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website and obtained as a result a **Pandas Dataframe** with all the relevant information for each course within these pages. Here is a brief description of each of the classes:

- `WebScraper`. This class performs web scraping on multiple pages of the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website. To do this it uses the following methods:

    - `scrape_urls()`. Scrapes the urls for all the courses within the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website and saves them in a text file.
    
    - `scrape_htmls()`. Scrapes and saves the HTML file for each URL contained in the text file produced by the `scrape_urls()` method.

- `HTMLParser`. This class parses each HTML file obtained with the `WebScraper.scrape_htmls()` method and obtains relevant information within these files. To do this it uses the following methods:

    - `parse_htmls()`. Parses information within the HTMLs (courses) obtained with the `WebScraper.scrape_htmls()` method and obtains: the name of the course, name of the university, name of the faculty, modality (MSc., full time/part time, online/on campus, etc.), description, start date, fees, duration, city, country, and URL, and saves all this information on tsv files.

    - `get_dataframe()`: Obtains a dataframe containing the parsed information of all the courses obtained with the `parse_htmls()` method.

For more information about the **implementation** of these classes and methods, please refer to their corresponding `web_scraper.py` and `html_parser.py` files contained in the `modules` directory of our repository.


### **1.1. Getting the list of Master's Degrees courses**

First, we start by web scraping the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website. In particular, we start by collecting the URLs associated with all the Master courses within the first 400 pages of this section, since each page has 15 courses, we obtained 6000 unique master's degree URLs. We obtained as an output a .txt file where each single line corresponds to a particular master's URL.

To do this, we used our `WebScraper` class. In particular, the `scrape_urls()` method of our `WebScraper` class:

In [2]:
#First, we have to initialize the WebScraper class by calling the constructor
web_scraper = WebScraper()

#As a second step, we can call the .scrape_urls() method to get the list of urls mentioned above. This method doesn't return anything. 
#Instead, the method saves the URLs in a text file called urls.txt
web_scraper.scrape_urls()


After performing the last step we generated an `urls.txt` file where each line was a different URL for a course within the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website. As an exercise, to see that our code indeed worked we can print the first line of this file:

In [2]:
#Here we print the first URL from the urls.txt file
with open('data/urls.txt', 'r') as f:
    print(f.readline())


https://www.findamasters.com/masters-degrees/course/3d-design-for-virtual-environments-msc/?i93d2645c19223



### **1.2. Crawling the master's degree pages**

Once we got the URLs of the courses contained in the first 400 pages of the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website, we saved the HTML files of each individual URL. We performed the following steps:

1. Downloading the HTML corresponding to each of the collected URLs.

2. Saving the downloaded `HTML` file into folders. Where each folder contains the URLs belonging to a given page. In other words, each folder contains the `HTML` files of the courses on page 1, page 2, etc. of the list of master's programs.
   
To do this, we used our `WebScraper` class. In particular, the `scrape_htmls()` method of our `WebScraper` class:

In [3]:
#Here, we can call the .scrape_htmls() method to get the HTMLs of the URLs mentioned above. This method doesn't return anything.
#Instead, the method saves the HTMLs in a folder called htmls, where this folder contains subfolders for each page for the first 400 pages.
#!!DO NOT RUN THIS METHOD IF YOU ALREADY HAVE THE HTMLS IN THE FOLDER!!
web_scraper.scrape_htmls()


Connection error: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')). Waiting 60 seconds and trying again.
Connection error: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response')). Waiting 60 seconds and trying again.


It is important to notice that this method takes to run approximately $\sim 10$ hours since we have to download $6000$ HTML files. We didn't try to optimize the running time since we encountered several issues that constrained our running time when performing web-scraping:

1. When accessing the information, we encountered an error: **Error 429: Too Many Requests** since we were requesting information **too fast**, this constrained us to take a given amount of time between each request in order to ensure the correct obtention of information.

3. We had a very *unstable* internet connection when trying to download the HTML files. In order to ensure the optimal obtention of data, we waited $60$ seconds every time there was a **ConnectionError** and tried again.

4. Finally, even though we took a given amount of time between each request, there were cases where we obtained HTML files with only the text *"Just a moment..."* written on it. Therefore we augmented the time threshold between each request in order to avoid this situation and clearly this took a big toll on the running time of our final code.

We consulted the following sources to solve each one of these problems:

- [Problem HTTP error 403 in Python 3 Web Scraping](https://stackoverflow.com/questions/16627227/problem-http-error-403-in-python-3-web-scraping).

- [urllib2 HTTP error 429](https://stackoverflow.com/questions/13213048/urllib2-http-error-429).

- [Web Scraping using Python (and Beautiful Soup)](https://www.datacamp.com/tutorial/web-scraping-using-python).

### **1.3 Parsing the downloaded pages**

At this point, we have saved all the HTML documents about the master's degree of interest. The next step we need to perform is extracting relevant information from each one of the courses we have. The **structure** of the web pages for each course can be exemplified with the following list/image:


<img align = "right" src="https://raw.githubusercontent.com/Sapienza-University-Rome/ADM/master/2023/Homework_3/img/example.jpeg" width="650" height = "450"/>

<div style="text-align: left"> 

The information we desire to obtain for each course and their format is as follows:

1. Course Name (<span style='color:#CB0900'> **CourseName** </span>): `string`.

2. University (<span style='color:#007B04'> **UniversityName** </span>): `string`.

3. Faculty (<span style='color:orange'> **facultyName** </span>): `string`.

4. Full or Part Time (<span style='color:#1A2E81'> **isItFullTime** </span>): `string`.

5. Short Description (<span style='color:black'> **description** </span>): `string`.

6. Start Date (<span style='color:purple'> **startDate** </span>): `string`.

7. Fees (<span style='color:pink'> **fees** </span>): `string`.

8. Modality (<span style='color:#FF9797'> **modality** </span>): `string`.

9. Duration (<span style='color:#3ABC3C'> **duration** </span>): `string`.

10. City (<span style='color:#F93345'> **city** </span>): `string`.

11. Country (<span style='color:yellow'> **country** </span>): `string`.

12. Presence or online modality (<span style='color:#0874FF'> **administration** </span>): `string`.

13. Link to the page (**url**): `string`.
</div>

To obtain this information, we used our `HTMLParser` class. In particular, the `parse_htmls()` method of our `HTMLParser` class. This method extracts the previous information from each HTML file of every course contained in the first 400 pages of the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website and saves it in a `.tsv` file. At the end we should finish with $6000$ `.tsv` files:

In [29]:
#First, we have to initialize the HTMLParser class by calling the constructor
html_parser = HTMLParser()

#As a second step, we can call the .parse_htmls() method to get the tsv files containing all the information mentioned before for each course.
#Instead, the method saves the URLs in a text file called urls.txt
html_parser.parse_htmls()


As an exercise, to see that our code indeed worked we can print the first `.tsv` file i.e., the `.tsv` for the first course in the first page of the [MSc. Degrees](https://www.findamasters.com/masters-degrees/msc-degrees/) section of the [Find a Masters](https://www.findamasters.com/) website:

In [3]:
#Here we print the first .tsv file from the data/tsvs folder
with open(f'data/tsvs/course_{1}.tsv', 'r') as f:
        print(f.read())


3D Design for Virtual Environments - MSc	Glasgow Caledonian University	School of Engineering and Built Environment	Full time	3D visualisation and animation play a role in many areas, and the popularity of these media just keeps growing. Digital animation provides the eye-catching special effects in the 21st century's favourite films and television shows; 3D design is also essential to everyday work in everything from computer games development, online virtual world development and industrial design to marketing, product design and architecture. GCU's programme in 3D Design for Virtual Environments will help you develop the skills to thrive in a successful career as a visual designer. The programme is practical and career-focused, oriented towards current industry needs, technology and practice. No prior knowledge of 3D design is required.	September	Please see the university website for further information on fees for this course.	MSc	1 year full-time	Glasgow	United Kingdom	On Campus	ht

Once we've obtained the relevant data we can finally visualize it properly by creating a *Pandas DataFrame*. To do this we can use the `get_dataframe()` method of our `HTMLParser` class. This method joins all the `.tsv` files created before in a single DataFrame:

In [2]:
#Here we can call the .get_dataframe() method to get the pandas dataframe containing all the information mentioned before for each course.
course_dataset = html_parser.get_dataframe()
#Here we print the first 5 rows of the dataframe
course_dataset.head()


courseName  \
0           3D Design for Virtual Environments - MSc   
1                       Accounting and Finance - MSc   
2  Accounting, Accountability & Financial Managem...   
3  Accounting, Financial Management and Digital B...   
4                                     Addictions MSc   

                  universityName  \
0  Glasgow Caledonian University   
1            University of Leeds   
2          King’s College London   
3          University of Reading   
4          King’s College London   

                                         facultyName isItFullTime  \
0        School of Engineering and Built Environment    Full time   
1                   Leeds University Business School    Full time   
2                             King’s Business School    Full time   
3                             Henley Business School    Full time   
4  Institute of Psychiatry, Psychology and Neuros...    Full time   

                                         description  startDate  \
0  3D visualisation and animation play a role in ...  September   
1  Businesses and governments rely on sound finan...  September   
2  Our Accounting, Accountability & Financial Man...  September   
3  Embark on a professional accounting career wit...  September   
4  Join us for an online session for prospective ...  September   

                                                fees modality  \
0  Please see the university website for further ...      MSc   
1  UK: £18,000 (Total) International: £34,750 (To...      MSc   
2  Please see the university website for further ...      MSc   
3  Please see the university website for further ...      MSc   
4  Please see the university website for further ...      MSc   

           duration     city         country administration  \
0  1 year full-time  Glasgow  United Kingdom      On Campus   
1  1 year full time    Leeds  United Kingdom      On Campus   
2         1 year FT   London  United Kingdom      On Campus   
3  1 year full time  Reading  United Kingdom      On Campus   
4       One year FT   London  United Kingdom      On Campus   

                                                 url  
0  https://www.findamasters.com/masters-degrees/c...  
1  https://www.findamasters.com/masters-degrees/c...  
2  https://www.findamasters.com/masters-degrees/c...  
3  https://www.findamasters.com/masters-degrees/c...  
4  https://www.findamasters.com/masters-degrees/c...

----

## **2. Search Engine**

Now that we have our data cleaned and organized in one dataset, we want to build a Search Engine that, given as input a query, return the courses that match the query. To do this we created **three** custom-made Python classes that helped us build our Search Engine and pre-process our data in order to make our Search Engine efficient and accurate. Here is a brief description of each of the classes:

- `DataPreprocesser`: This class preprocesses the fees column and the text columns (removes punctuation/stopwords and performs tokenizing and lemmatizing) of the MSc courses dataset. To do this it uses the following methods:

    - `preprocess_fees_column()`. Finds and converts the course fees into Euros, obtains the maximum fee and saves it as a float in a new column.

    - `preprocess_text_column()`. Removes punctuation and stopwords from the text, and performs tokenizing and lemmatizing of every word.

- `SearchEngine`:  This class implements a Search Engine that queries documents by their description and returns the documents that include **all** the words in the given query. To do this it uses the following method:

    - `query()`. Function that queries the dataset and returns the documents that contain all the words in a given text query sorted by index.

* `TopKSearchEngine`: This class implements a Search Engine that queries documents by their description and returns the documents that include **all** the words in the given query sorted by their **cosine similarity** with the query. This class inherits from the `SearchEngine` parent class and has the same method:

    - `query()`. Function that queries the dataset and returns the documents that contain all the words in a given text query sorted by cosine similarity. 

For more information about the **implementation** of these classes and methods, please refer to their corresponding `data_preprocesser.py` and `search_engine.py` files contained in the `modules` directory of our repository.

### **2.0. Preprocessing the text**

First, we must pre-process the text information collected for each MSc in order to make our Search Engine more accurate in finding documents. To do this we performed the following data-processing pipeline:

1. First, we performed text **tokenization**. Text can be thought as being formed by units called *tokens* that build up the entire text corpus and the process of splitting a text sequence into its tokens is *tokenization*. For example, when tokenizing the string *"Hello, World!"* we find four tokens: *"hello"* *","* *"world"* *"!"*.

    As we can see, while tokenizing we also take into account punctuation as tokens. This is why, we chose to also **remove punctuation** when performing tokenization. Another important thing to notice is that we chose to convert all our text to lower case in other to treat words like "Hello" and "hello" as the same.

2. Once we tokenized the text, we then removed **stopwords**. Stopwords are a set of commonly used words in a language. Examples of stop words in English are “a,” “the,” “is,” “are,” etc. This words don't carry much information and they are not important when performing a query, this is why we chose to eliminate them along with other **non-alphanumeric** characters.

3. Finally, once we tokenized and removed the stopwords we performed **lemmatization** on the text. Lemmatization is another technique used to reduce inflected words to their root word. It describes the algorithmic process of identifying an inflected word’s *lemma* (dictionary form) based on its intended meaning. For example, when lemmatizing the word *"running"*, we obtain its lemma: *"run"*.

    We chose lemmatization as opposed to **stemming** since stemming algorithms function by taking a list of frequent prefixes and suffixes found in inflected words and chopping off the end or beginning of the word. This can occasionally result in word stems that are not real words. Lemmatization is more accurate and takes into account the **meaning** of the word.

In the end, we ended up with tokenized text containing words reduced to their roots that synthesize the content of each document. As an example of our process, let's pre-process the string: *"I like eating Carbonara. Carbonara is a type of pasta."*:

1. **Tokens (lowercase and without punctuation)**: *"i"* *"like"* *"eating"* *"carbonara"* *"carbonara"* *"is"* *"a"* *"type"* *"of"* *"pasta"*

2. **Tokens (without stop words)**: *"like"* *"eating"* *"carbonara"* *"carbonara"* *"type"* *"pasta"*

3. **Lemmatized tokens**: *"like"* *"eat"* *"carbonara"* *"carbonara"* *"type"* *"pasta"*

As we can see, we end up with key words without giving up the main idea of the text. We performed this text pre-processing for the following text fields of our dataset since they will be useful throughout this notebook:

- `description`: A brief description of the Master's Degree course.

- `courseName`: The name of the course.

- `universityName`: The name of the university.

- `facultyName`: The faculty name.

- `city`: The city where the course takes place.

- `country`: The country where the course takes place.

To do this, we used our `DataPreprocesser` class. In particular, the `preprocess_text_column()` method of our `DataPreprocesser` class:

In [3]:
#Here we initialize the DataPreprocesser class by calling the constructor
data_preprocesser = DataPreprocesser(course_dataset)

#Here we call the .preprocess_text_column() method to preprocess the description column of the dataframe
data_preprocesser.preprocess_text_column(column_name="description")
#Here we call the .preprocess_text_column() method to preprocess the courseName column of the dataframe
data_preprocesser.preprocess_text_column(column_name="courseName")
#Here we call the .preprocess_text_column() method to preprocess the universityName column of the dataframe
data_preprocesser.preprocess_text_column(column_name="universityName")
#Here we call the .preprocess_text_column() method to preprocess the facultyName column of the dataframe
data_preprocesser.preprocess_text_column(column_name="facultyName")
#Here we call the .preprocess_text_column() method to preprocess the city column of the dataframe
data_preprocesser.preprocess_text_column(column_name="city")
#Here we call the .preprocess_text_column() method to preprocess the country column of the dataframe
data_preprocesser.preprocess_text_column(column_name="country")


We can observe the resulting preprocessed columns for our dataset:

In [4]:
#Here we print the first 5 rows of the dataframe with the preprocessed columns
data_preprocesser.dataset[["description (PROCESSED)", "courseName (PROCESSED)", "universityName (PROCESSED)", "facultyName (PROCESSED)", "city (PROCESSED)", "country (PROCESSED)"]].head()


description (PROCESSED)  \
0  3d,visualisation,animation,play,role,many,area...   
1  business,government,rely,sound,financial,knowl...   
2  accounting,accountability,financial,management...   
3  embark,professional,accounting,career,academic...   
4  join,u,online,session,prospective,student,find...   

                              courseName (PROCESSED)  \
0                  3d,design,virtual,environment,msc   
1                             accounting,finance,msc   
2  accounting,accountability,financial,management...   
3  account,financial,management,digital,business,msc   
4                                      addiction,msc   

      universityName (PROCESSED)  \
0  glasgow,caledonian,university   
1               university,leeds   
2            king,college,london   
3             university,reading   
4            king,college,london   

                        facultyName (PROCESSED) city (PROCESSED)  \
0          school,engineering,build,environment          glasgow   
1              leeds,university,business,school            leeds   
2                          king,business,school           london   
3                        henley,business,school          reading   
4  institute,psychiatry,psychology,neuroscience           london   

  country (PROCESSED)  
0      united,kingdom  
1      united,kingdom  
2      united,kingdom  
3      united,kingdom  
4      united,kingdom

### **2.0.1 Preprocessing the fees column**

We also want the preprocess the ```fees``` field in order to collect numeric information. To do this we followed these steps:

1. First, we found all the currency strings included in the ```fees```. Since there are a lot of different ways to write textual currency we checked the most important cases using Regular Expressions (for more information, check the `data_preprocesser.py` module). In particular we extracted strings with the following formats:

    * `(\p{Sc})\s?(\d+[\.\,\s]{0,1}[\d\.\,]{0,}`: Strings with format € 2,209, $2,209, etc.

    * `(\d+[\.\,\s]{0,1}[\d\.\,]{0,})\s?(\p{Sc})`: Strings with format 2,209 €, 2,209$, etc.
    
    * `([A-Z\p{Sc}]{3})\s?(\d+[\.\,\s]{0,1}[\d\.\,]{0,})`: Strings with format EUR 2,209, USD2,209, etc.

    * `(\d+[\.\,\s]{0,1}[\d\.\,]{0,})\s?([A-Z\p{Sc}]{3})`: Strings with format 2,209 EUR, 2,209USD, etc.

    * `\p{Sc}\s?(\d+[\.\,\s]{0,1}[\d\.\,]{0,})\s?([A-Z]{3})`: Strings with format € 2,209 EUR, $2,209USD, etc.

    * `([A-Z]{3})\s?\p{Sc}\s?(\d+[\.\,\s]{0,1}[\d\.\,]{0,})`: Strings with format EUR € 2,209, USD$2,209, etc.

    2. Once we found all the currency strings, we grouped them in tuples `(value, currency)` in order to distinguish the symbols/ISO codes of the currencies from the actual numeric values. At the same time, we eliminated duplicates since we could have the case that the same value is written in two different ways. For example: `(200, USD)` and `(200, $)`. We made sure to keep the currency [ISO](https://en.wikipedia.org/wiki/ISO_4217) code when present since it is helpful when trying to convert currencies.

    3. Once we had all the numeric values without duplicates, we proceeded to convert all the currencies to a common currency (in this case, EUR). To do this we first asked ChatGPT for help in finding an API that could help us performing this task. The prompt we wrote was:

        > I have a DataFrame column containing values in different currencies. How can I convert them to euros using Python?

        And the response was:

        > To convert values in different currencies to euros in a DataFrame column using Python, you can use a currency conversion API or a library that provides exchange rates. One popular library for this task is the `forex-python` library. First, you need to install the library if you haven't already:
        >```bash
        > pip install forex-python
        >```
        > Then, you can use the library to convert currencies. Here's an example:
        >```python
        >import pandas as pd
        >from forex_python.converter import CurrencyRates
        >
        ># Create a sample DataFrame
        >data = {'Amount': [100, 150, 200],
        >        'Currency': ['USD', 'GBP', 'JPY']}
        >df = pd.DataFrame(data)
        >
        ># Function to convert currency to euros
        >def convert_to_euros(row):
        >    currency_rates = CurrencyRates()
        >    exchange_rate = currency_rates.get_rate(row['Currency'], 'EUR')
        >    return row['Amount'] * exchange_rate
        >
        ># Apply the conversion function to the DataFrame
        >df['Amount_in_EUR'] = df.apply(convert_to_euros, axis=1)
        >
        ># Display the resulting DataFrame
        >print(df)
        >
        >```
        >In this example, the convert_to_euros function takes a row of the DataFrame, retrieves the exchange rate for the given currency against euros using the `forex-python` library, and then calculates the equivalent amount in euros. The apply method is used to apply this function to each row of the DataFrame, and a new column 'Amount_in_EUR' is created to store the converted values.
        >
        >Note that you need an internet connection to fetch the exchange rates from the API. If you have a large DataFrame, you may want to explore more efficient ways to perform the conversion, such as parallel processing or vectorized operations.

        We used the `forex-python` library to convert all the fees we found to EUR by mapping each currency to its corresponding ISO code and then performing the conversion. In order to avoid not finding conversion rates, we always took the conversion rate of the day before we coded this part.

    4. Finally, once we had all the fees in EUR we performed another step. When there were more than one fees in a given Master's course, we took the highest fee as our fee and ignored the others. In the end, we finished with a column filled with only a numeric value in EUR (or NA if there was no fee).

To do this, we used our `DataPreprocesser` class. In particular, the `preprocess_fees_column()` method of our `DataPreprocesser` class:

In [5]:
#Here we call the .preprocess_fees_column() method to preprocess the fees column of the dataframe
data_preprocesser.preprocess_fees_column()


We can observe the resulting preprocessed fees column for our dataset:

In [6]:
#Here we print the first 5 rows of the processed fees column
data_preprocesser.dataset[["fees", "fees (EUR)"]].head()


fees  fees (EUR)
0  Please see the university website for further ...         NaN
1  UK: £18,000 (Total) International: £34,750 (To...   39871.493
2  Please see the university website for further ...         NaN
3  Please see the university website for further ...         NaN
4  Please see the university website for further ...         NaN

As we can see, within the first 5 rows of our dataset, only the second row contained information for fees, resulting in a fee numeric value of $39871$ EUR.

### **2.1. Conjunctive query**

For the first version of our Search Engine, we narrowed our interest to the `description` column of each course. This means that our Search Engine will evaluate queries only concerning the course's description. To evaluate the queries our Search Engine follows the following steps:

1. It builds a dictionary containing all the (preprocessed) words found in the `description` column of each course. Using this vocabulary it built the **inverted index** of the words. The inverted index is a dictionary of the form:
    ```
    {
    term_id_1:[document_1, document_2, document_4],
    term_id_2:[document_1, document_3, document_5, document_6],
    ...}
    ```
    where  `document_i` is the *id* of a document that contains a specific word. The `term_id_i` is the *id* of a specific word in the vocabulary.

    This index gives a mapping from every word in the vocabulary to all the documents that contain it.

2. Once it built (and saved) the vocabulary and inverted index, it takes a **query** as an input. As a first step it preprocesses the query and then searchs for all the documents that contain **all** of the words/tokens in the query. To perform this search it takes advantage of the inverted index since it only has to take the intersection of the lists of the terms contained in the query.

3. Once it gets all the relevant documents, it returns them sorted by index. The sorting part was optional but we decided to do it since it is more visually appealing.

### **2.1.1 Creating your index!**

As we mentioned before, the vocabulary and the inverted index are the building blocks of our Search Engine, this is why they are obtained *before* making any query to the Search Engine and saved into memory. In this way they are loaded into memory when necessary instead of being calculated each time. To do this we incorporated them as class **attributes** so they are loaded into memory each time the `SearchEngine` class is initialized.

As an exercise, we can see the structure of the vocabulary:



In [7]:
#Here we load the vocabulary of the description column
with open(f"data/description_vocabulary.pickle", "rb") as f:
                vocabulary = pickle.load(f)

#As an example we can print the index of the word "data"
print(vocabulary["data"])


2193


As we can observe, the `term_id` of the word *data* is $2193$. Therefore if we search this id within the inverted index:

In [8]:
#Here we load the inverted index of the description column
with open(f"data/description_inverted_index.pickle", "rb") as f:
                inverted_index = pickle.load(f)

#As an example we can print the inverted index of the word "data" (at least the first 10 documents)
print(inverted_index[2193][:10])


[13, 19, 20, 24, 34, 35, 36, 57, 58, 60]


We can observe that the first 10 documents that contain the word *data*  in their description field are the documents: 13, 19, 20, 24, 34, 35, 36, 57, 58 and 60.

#### **2.1.2 Execute the query**

We can test our Search Engine by executing the query "advanced knowledge". To do this, we use our `SearchEngine` class. In particular, the `query()` method of our `SearchEngine` class:

In [10]:
#First we initialize the SearchEngine class by calling the constructor. In this step is when the vocabulary and the inverted index are created/saved or loaded into memory
search_engine = SearchEngine(data_preprocesser.dataset)

#Here we call the .query() method to get the results of the query "advanced knowledge"
result = search_engine.query("advanced knowledge")
#Here we print the first 5 results
result.head()


courseName  \
1                   Accounting and Finance - MSc   
198         Master of Public Health (MPH) Online   
204  Master of Science in Mechanical Engineering   
215      Materials Science and Engineering - MSc   
239                MSc Advanced Computer Science   

                                        universityName  \
1                                  University of Leeds   
198                           Brunel University Online   
204  The Hong Kong University of Science and Techno...   
215                                University of Leeds   
239                            University of Sheffield   

                                           description  \
1    Businesses and governments rely on sound finan...   
198  Brunel’s Master of Public Health (online) has ...   
204  The program is designed to benefit students wi...   
215  Materials science is at the forefront of provi...   
239  This MSc keeps you at the cutting edge of deve...   

                                                   url  
1    https://www.findamasters.com/masters-degrees/c...  
198  https://www.findamasters.com/masters-degrees/c...  
204  https://www.findamasters.com/masters-degrees/c...  
215  https://www.findamasters.com/masters-degrees/c...  
239  https://www.findamasters.com/masters-degrees/c...

We only printed the first 5 documents in order to keep the Notebook as tidy as possible. Nevertheless we can see that the total number of documents that contain all the words in our query are 188. This is a big quantity and we can't distinguish one from another, this is why it is necessary to make some adjustments to this basic Search Engine.

In [12]:
#Here we print the number of results that the query "advanced knowledge" returned
print(f"Number of results: {len(result)}")


Number of results: 188


### **2.2 Conjunctive query & Ranking score**

For the second Search Engine, given a query, we want to get the *top-k* (in this case, we chose $k=5$) documents related to the query and a **similarity** measure. In particular we chose to perform the following procedure:

1. We built a dictionary containing all the words found in the `description` column of each course (in reality, use the one we created before). Using this vocabulary we built the **TfIdf inverted index** of the words. The TfIdf inverted index is a dictionary of the form:
    ```
    {
    term_id_1:[(document_1, tfIdf_{term,document1}), (document_2, tfIdf_{term,document2}), (document_4, tfIdf_{term,document4})],
    term_id_2:[(document_1, tfIdf_{term,document1}), (document_3, tfIdf_{term,document3}), (document_5, tfIdf_{term,document5}), (document_6, tfIdf_{term,document6})],
    ...}
    ```
    where `document_i` is the *id* of a document that contains a specific word, the `term_id_i` is the *id* of a specific word in the vocabulary, and the `TfIdf` is the Term Frequency-Inverse Document Frequency value of the `term_id_i` within `document_i`. The Term Frequency-Inverse Document Frequency is given by:

    \begin{equation}
    \text{TfIdf}(t,d) = \text{Tf}(t,d) \times \text{Idf}(t) =  \text{Tf}(t,d) \times \log \left(\frac{N}{1+df}\right),
    \end{equation}

    where $\text{Tf}(t,d)$ is the (normalized) frequency of times the term $t$ appears in document $d$ and $\text{Idf}(t)$ is the **inverse document frequency** of term $t$, where $df$ is the number of documents that include term $t$ and $N$ is the total number of documents. The purpose of the Idf is to give higher weight to terms that are rare across the entire corpus of a text and lower weight to terms that are common.

    The TfIdf index gives a mapping from every word in the vocabulary to all the documents that contain it and a measurement **how important** is that word within each document relative to its importance across all documents. 

2. Once we built (and saved) the vocabulary and TfIdf inverted index, we take a **query** as an input. As a first step we preprocess the query and then search for all the documents that contain **all** of the words/tokens in the query. To perform this search we takes advantage of the TfIdf inverted index since we only have to take the intersection of the lists of the terms contained in the query (the lists of the first elements of the tuple).

3. Once we got all the relevant documents, we decided to sort them by their **Cosine Similarity** with respect to the query. The Cosine Similarity is a vector similarity measure that measures how similar are two vectors $\vec{A}$ and $\vec{B}$ by taking the angle $\theta$ between them. It is given by:

    \begin{equation}
    \text{sim}\left(\vec{A}, \vec{B}\right) = cos(\theta) = \frac{\vec{A}\cdot\vec{B}}{|\vec{A}||\vec{B}|}.
    \tag{2}
    \end{equation}

    In the context of NLP, we can represent documents as vectors where each vector value is the tfIdf representation of a term within this document. Therefore, we can obtain the similarity between documents by taking the **cosine similarity** of their tfIdf representation. In this case, we obtained the cosine similarity between the query and all the obtained documents. 
    
    Once we had a similarity value for all the relevant documents, we **sorted** them in descending value with respect to the similarity with the query. In order to mantain the top-$k$ documents in an efficient way, we used a **max-heap** data structure.

#### **2.2.1 Inverted index**

As in the first Search Engine we built, the TfIdf inverted index is a fundamental tool for building of our Search Engine, this is why they are obtained *before* making any query to the Search Engine and saved into memory. In this way they are loaded into memory when necessary instead of being calculated each time. To do this we incorporated them as a class **attribute** so it was loaded into memory each time the `TopKSearchEngine` class is initialized.

As an exercise, we can observe the first 5 elements of the TfIdf inverted index value for the "data" term as we did before:

In [17]:
#Here we load the inverted index of the description column
with open(f"data/description_tfIdf_inverted_index.pickle", "rb") as f:
                tf_idf_inverted_index = pickle.load(f)

#As an example we can print the inverted index of the word "data" (at least the first 5 documents)
print(tf_idf_inverted_index[2193][:5])


[(13, 0.08754502045096446), (19, 0.10059505738094911), (20, 0.09754119920675935), (24, 0.054922388160716075), (34, 0.21374332346270045)]


#### **2.2.2 Execute the query**

We can test our Search Engine by executing once again the query "advanced knowledge". To do this, we use our `TopKSearchEngine` class. In particular, the `query()` method of our `TopKSearchEngine` class:

In [18]:
#First we initialize the TopKSearchEngine class by calling the constructor. In this step is when the vocabulary and the inverted index are created/saved or loaded into memory
search_engine = TopKSearchEngine(data_preprocesser.dataset)
#Here we call the .query() method to get the results of the query "advanced knowledge"
result = search_engine.query("advanced knowledge")
#Here we print the first 5 results
result


courseName  \
651                    Advanced Clinical Practice - MSc   
5242  Management and Digital Business (with Advanced...   
753                              Advanced Computing MSc   
857   Advanced Nurse Practitioner/Professional Pract...   
698                      Advanced Clinical Practice MSc   

                               universityName  \
651       Canterbury Christ Church University   
5242         Liverpool John Moores University   
753                     King’s College London   
857   University of the Highlands and Islands   
698                   University of Greenwich   

                                            description  \
651   Gain the knowledge and skills needed to become...   
5242  This Advanced Practice course provides an in-d...   
753   Our Advanced Computing MSc provides knowledge ...   
857   Developed in partnership with expert clinical ...   
698   Develop your skills and deepen your knowledge ...   

                                                    url  similarity  
651   https://www.findamasters.com/masters-degrees/c...    0.343794  
5242  https://www.findamasters.com/masters-degrees/c...    0.341632  
753   https://www.findamasters.com/masters-degrees/c...    0.335160  
857   https://www.findamasters.com/masters-degrees/c...    0.303509  
698   https://www.findamasters.com/masters-degrees/c...    0.296750

----

## 3. **Define a new score!**

Can we do better? In our last implementation of our Search Engine we sorted our results by the **cosine similarity** of the TfIdf representation of their description column with respect to the TfIdf representation of the query. Nevertheless this doesn't take into account the following things:

1. A particular course can be **more** similar to a particular query if it includes the query words **also** in other information fields. For example:

     If we search for the query *"engineering"* we obtain a similarity measure of the results regarding the frequency of this word within the description of each document but we **ignore** if this words also appear say in the `courseName` or even the `facultyName` of the course. In many cases this would be the case since generally engineering courses are given by engineering schools and have engineering names.

2. Users are generally more interested in finding courses in **places** they are interested in (cities, countries or even universities). Therefore our Search Engine could favour queries that include this data within their text. For example if we search "engineering Rome", we should define a similarity measure that favours engineering courses that indeed are taken in Rome.

To this end we propose a **weighted** cosine similarity measure, that takes into account not only the `description` field of our documents, but also other main information fields (`courseName`, `facultyName`) and location fields (`universityName`, `country`, `city`). Our new similarity measure is defined as:

\begin{equation}
\text{similarity}(\vec{d}, \vec{q}) = \frac{w_{\text{context}}}{3}\cdot \left[\text{cs}_{\text{description}}(\vec{d}, \vec{q}) + \text{cs}_{\text{course}}(\vec{d}, \vec{q}) + \text{cs}_{\text{faculty}}(\vec{d}, \vec{q})\right] + \frac{w_{\text{places}}}{3}\cdot \left[\text{cs}_{\text{university}}(\vec{d}, \vec{q}) + \text{cs}_{\text{country}}(\vec{d}, \vec{q}) + \text{cs}_{\text{city}}(\vec{d}, \vec{q})\right],
\tag{3}
\end{equation}

where $\vec{d}$ and $\vec{q}$ are the TfIdf representations of a given document and a query and the $1/3$ are normalization factors to keep the measure between 0 and 1. Let's explain this measure a little better:

The purpose of the similarity measure defined in $(3)$ is give more (or less) weight to **main information** related characteristics (description, course name and faculty name) of a given document i.e. we want our similarity measure to take into account words that relate to the general information of a given course. In the same way, we want to give more (or less) weight to **location** related characteristics (university, country and city) in case the user wants to find courses in a particular place around the world.

The choice of the weights is up to us and it depends on which characteristics we deem more relevant. In our case we decided to give more weight to the **location** related characteristics since as international students, finding Masters courses around the world is something we're very familiar with and we consider very important. This takes us to the final version of our equation:

\begin{equation}
\text{similarity}(\vec{d}, \vec{q}) = \frac{1}{9}\cdot \left[\text{cs}_{\text{description}}(\vec{d}, \vec{q}) + \text{cs}_{\text{course}}(\vec{d}, \vec{q}) + \text{cs}_{\text{faculty}}(\vec{d}, \vec{q})\right] + \frac{2}{9}\cdot \left[\text{cs}_{\text{university}}(\vec{d}, \vec{q}) + \text{cs}_{\text{country}}(\vec{d}, \vec{q}) + \text{cs}_{\text{city}}(\vec{d}, \vec{q})\right],
\tag{4}
\end{equation}

where we defined $w_{\text{context}} = 1/3$ and $w_{\text{places}} = 2/3$. We can see that this measure favours location-related queries with respect to information-related queries and in case a word appears only in **one** column, this reduces to a scaled version of the original **cosine similarity** we used before.

>**Note:** Since we are only defining a **new** similarity measure it is important to account that we're working with the vocabulary we built for the `description` column, therefore if a word in the query **doesn't** appear in the description field it doesn't matter if it is on the other fields, the Search Engine won't find a match. The purpose of the weighted similarity measure is to favour queries that appear in the description field **and** in other fields as well.

### **3.1 Executing a location-based query**

In order to test our brand new similarity measure, we created a custom-made Python class called `WeightedTopKSearchEngine`. This class implements a Search Engine that queries documents by their description and returns the documents that include **all** the words in the given query sorted by their **weighted cosine similarity** (defined above) with the query. This class inherits from the `TopKSearchEngine` parent class and has the same method:

- `query()`. Function that queries the dataset and returns the documents that contain all the words in a given text query sorted by their weighted cosine similarity. 

For more information about the **implementation** of this class and method, please refer to the corresponding `search_engine.py` file contained in the `modules` directory of our repository.

First, we can test our Search Engine by executing a location-based query: "engineering London". If our assumptions are correct, this `WeightedTopKSearchEngine` should sort in a better way the documents than its "simpler" version, the `TopKSearchEngine`. To do this, we can use the `query()` method of our `WeightedTopKSearchEngine` class:

In [10]:
#Here we initialize the WeightedTopKSearchEngine class by calling the constructor. 
search_engine = WeightedTopKSearchEngine(data_preprocesser.dataset)
#Here we call the .query() method to get the results of the query "engineering London"
result = search_engine.query("engineering London")
#Here we print the first 5 results
result


courseName  \
1928                              Civil Engineering MSc   
4085                  Global Innovation Design - MA/MSc   
4555               Innovation Design Engineering MA/MSc   
5089  Lightweight Structures and Impact Engineering MSc   
2927              Disability, Design and Innovation MSc   

                 universityName  \
1928  University of East London   
4085    Imperial College London   
4555       Royal College of Art   
5089   Brunel University London   
2927  University College London   

                                            description  \
1928  Our MSc Civil Engineering degree will develop ...   
4085  Global Innovation Design is a joint Master's d...   
4555  The MA/ MSc Innovation Design Engineering (IDE...   
5089  NSIRC industry scholarships are available. Del...   
2927  Become a pioneer in disability innovation and ...   

                                                    url  similarity  
1928  https://www.findamasters.com/masters-degrees/c...    0.220554  
4085  https://www.findamasters.com/masters-degrees/c...    0.172554  
4555  https://www.findamasters.com/masters-degrees/c...    0.156729  
5089  https://www.findamasters.com/masters-degrees/c...    0.148807  
2927  https://www.findamasters.com/masters-degrees/c...    0.146027

If we compare the results of this query with the same query applied to the `TopKSearchEngine` class, we can see that:

In [11]:
#Here we initialize the WeightedTopKSearchEngine class by calling the constructor. 
search_engine = TopKSearchEngine(data_preprocesser.dataset)
#Here we call the .query() method to get the results of the query "engineering London"
result = search_engine.query("engineering London")
#Here we print the first 5 results
result


courseName  \
1928                             Civil Engineering MSc   
4555              Innovation Design Engineering MA/MSc   
2927             Disability, Design and Innovation MSc   
5184  Machine Learning and Data Science (Online) - MSc   
4085                 Global Innovation Design - MA/MSc   

                 universityName  \
1928  University of East London   
4555       Royal College of Art   
2927  University College London   
5184    Imperial College London   
4085    Imperial College London   

                                            description  \
1928  Our MSc Civil Engineering degree will develop ...   
4555  The MA/ MSc Innovation Design Engineering (IDE...   
2927  Become a pioneer in disability innovation and ...   
5184  This Master's course aims to accelerate your c...   
4085  Global Innovation Design is a joint Master's d...   

                                                    url  similarity  
1928  https://www.findamasters.com/masters-degrees/c...    0.324583  
4555  https://www.findamasters.com/masters-degrees/c...    0.295204  
2927  https://www.findamasters.com/masters-degrees/c...    0.260738  
5184  https://www.findamasters.com/masters-degrees/c...    0.161661  
4085  https://www.findamasters.com/masters-degrees/c...    0.147866

From this we can conclude the following:

- Similarity values are generally **lower** for our custom similarity measure than the original cosine similarity. This is expected since we're taking into account more parameters and therefore if a word doesn't exist in **all** of them, values will be gradually lower.

- Our similarity approach indeed favours courses where locations are mentioned throughout the name, faculty, description, etc. As we can see, the Search Engine with our custom-made similarity almost always returns courses where the name has *"engineering"* in it and the university has the word *"London"* in it. Nevertheless we can see that results are not that far away for both Search Engines, suggesting that maybe we could just stick with the original similarity approach and we would obtain similar results.

### **3.2 Executing a non-location-based query**

Now, we can test the other way around, a query that doesn't include location information and just wants to find a course by some keywords. We could use our previous query: "advanced knowledge" and see what we obtain:

In [12]:
#Here we initialize the WeightedTopKSearchEngine class by calling the constructor. 
search_engine = WeightedTopKSearchEngine(data_preprocesser.dataset)
#Here we call the .query() method to get the results of the query "advanced knowledge"
result = search_engine.query("advanced knowledge")
#Here we print the first 5 results
result


courseName  \
712  Advanced Clinical Practitioner Master's Degree...   
651                   Advanced Clinical Practice - MSc   
753                             Advanced Computing MSc   
698                     Advanced Clinical Practice MSc   
857  Advanced Nurse Practitioner/Professional Pract...   

                              universityName  \
712                 Brunel University London   
651      Canterbury Christ Church University   
753                    King’s College London   
698                  University of Greenwich   
857  University of the Highlands and Islands   

                                           description  \
712  The Advanced Clinical Practice Apprenticeship ...   
651  Gain the knowledge and skills needed to become...   
753  Our Advanced Computing MSc provides knowledge ...   
698  Develop your skills and deepen your knowledge ...   
857  Developed in partnership with expert clinical ...   

                                                   url  similarity  
712  https://www.findamasters.com/masters-degrees/c...    0.076331  
651  https://www.findamasters.com/masters-degrees/c...    0.069608  
753  https://www.findamasters.com/masters-degrees/c...    0.069516  
698  https://www.findamasters.com/masters-degrees/c...    0.064381  
857  https://www.findamasters.com/masters-degrees/c...    0.058131

Now for our benchmark Search Engine:

In [13]:
#Here we initialize the WeightedTopKSearchEngine class by calling the constructor. 
search_engine = TopKSearchEngine(data_preprocesser.dataset)
#Here we call the .query() method to get the results of the query "advanced knowledge"
result = search_engine.query("advanced knowledge")
#Here we print the first 5 results
result


courseName  \
651                    Advanced Clinical Practice - MSc   
5242  Management and Digital Business (with Advanced...   
753                              Advanced Computing MSc   
857   Advanced Nurse Practitioner/Professional Pract...   
698                      Advanced Clinical Practice MSc   

                               universityName  \
651       Canterbury Christ Church University   
5242         Liverpool John Moores University   
753                     King’s College London   
857   University of the Highlands and Islands   
698                   University of Greenwich   

                                            description  \
651   Gain the knowledge and skills needed to become...   
5242  This Advanced Practice course provides an in-d...   
753   Our Advanced Computing MSc provides knowledge ...   
857   Developed in partnership with expert clinical ...   
698   Develop your skills and deepen your knowledge ...   

                                                    url  similarity  
651   https://www.findamasters.com/masters-degrees/c...    0.343794  
5242  https://www.findamasters.com/masters-degrees/c...    0.341632  
753   https://www.findamasters.com/masters-degrees/c...    0.335160  
857   https://www.findamasters.com/masters-degrees/c...    0.303509  
698   https://www.findamasters.com/masters-degrees/c...    0.296750

From this we can observe the following:

- Similarity values **decrease** in an important way. We assume that since "advanced knowledge" is a query that cannot be generally found in location-based fields of our document, our similarity measure reflects this by giving very low similarity values for this given query.


- Our similarity approach returns documents that contain also "advanced knowledge" within their names, university names, etc. In this case this is not really convenient since we would want to find courses that provide us with "advanced knowledge" and not courses that contain "advanced" in other fields of the document. In conclusions, context is important for this approach to work well.

We can generally conclude that our approach presents us with **better** results when querying locations or specific information details like university names. Nevertheless it has a lot of weakness when providing results for general queries like we showed above. Although we cannot say our approach is better than the original approach, we can say that it works better in specific contexts.

----

## **4. Visualizing the most relevant MSc degrees**

Using maps can help people understand how far one university is from another so they can plan their academic careers more adequately. Therefore in this part we were asked to show a map of the courses found with our score defined in **Section 3**. To do this we performed the following process:

1. Once we obtained the results from our Search Engine we added extra-information of these courses we wanted to show on our plot to our resulting DataFrame. This extra information is shown when **hovering** over a point (course) in our map and includes:

    - `courseName`
    - `universityName`
    - `facultyName`
    - `fees`
    
    - `duration`

2. Along with this information we obtained the **geographical** information (longitude and latitude) of each course within our results. To do this we performed the following steps:

    - We built a new column called `fullAddress` made of the sum of the `facultyName`, `universityName`, `city` and `country` columns. For example, a value of this column could be: "Engineering School, London University, London, United Kingdom".

    - Once we've built this new column we proceeded to obtain the longitude and latitude of each `fullAddress` value by calling the Google Maps API and obtaining this information. We stored it in new columns called `lat` and `long`.

    - Finally we used all this information to build a map using the Plotly Express package.

To do this we built a custom-made class called `Map Plotter`. This class plots a map containing the most similar courses found with the SearchEngine class. It do this using the following method:

- `plot()`: Function that plots the map with the most similar courses found.

For more information about the **implementation** of this class and method, please refer to their corresponding `map_plotter.py` file contained in the `modules` directory of our repository. We can make an example using the results of our Search Engine that sorts results by their **weighted** cosine similarity. Using the results obtained before:
    


In [7]:
#Here we initialize the WeightedTopKSearchEngine class by calling the constructor. 
search_engine = WeightedTopKSearchEngine(data_preprocesser.dataset)
#Here we call the .query() method to get the results of the query "engineering London"
result = search_engine.query("engineering London")
#Here we print the first 5 results
result


courseName  \
1928                              Civil Engineering MSc   
4085                  Global Innovation Design - MA/MSc   
4555               Innovation Design Engineering MA/MSc   
5089  Lightweight Structures and Impact Engineering MSc   
2927              Disability, Design and Innovation MSc   

                 universityName  \
1928  University of East London   
4085    Imperial College London   
4555       Royal College of Art   
5089   Brunel University London   
2927  University College London   

                                            description  \
1928  Our MSc Civil Engineering degree will develop ...   
4085  Global Innovation Design is a joint Master's d...   
4555  The MA/ MSc Innovation Design Engineering (IDE...   
5089  NSIRC industry scholarships are available. Del...   
2927  Become a pioneer in disability innovation and ...   

                                                    url  similarity  
1928  https://www.findamasters.com/masters-degrees/c...    0.220554  
4085  https://www.findamasters.com/masters-degrees/c...    0.172554  
4555  https://www.findamasters.com/masters-degrees/c...    0.156729  
5089  https://www.findamasters.com/masters-degrees/c...    0.148807  
2927  https://www.findamasters.com/masters-degrees/c...    0.146027

We obtain the following map:

In [8]:
#Here we initialize the MapPlotter class by calling the constructor.
map_plotter = MapPlotter(data_preprocesser.dataset)
#Here we call the .plot() method to plot the results of the query "engineering London"
map_plotter.plot(result)


As we can see, this is an interactive map that shows the top-5 courses resulting from our query. This plot has the following features:

- You can zoom in/out as long as you want in case not all the courses are found in the same country. It also has street resolution in case we want to observe the exact location where our course takes place.

- The **size** of any point in the map represents the **weighted** cosine similarity of the course with our query in order to highlight courses that are more similar to what we wanted to find.

- The **color** of any point in the map represents the **fees** (in euros) of the course as we can see by the colorbar. In case a fee isn't included in the course information or we didn't convert it to euros, it has a default gray color.

- When hovering over a point one can obtain basic information about the course including: name, faculty, university, duration, etc.

----

## **6. Command Line Question**

For this Command Line question we were asked to perform the following things:

1. First, take the `course_i.tsv` files we created in **Section 1** and merge them using Linux commands.

2. Obtain the following information from our new `merged_courses.tsv` file:

    * Which country offers the most Master's Degrees? Which city?
    * How many colleges offer Part-Time education?
    * Print the percentage of courses in Engineering (the word "Engineer" is contained in the course's name).

To solve this problems we wrote an executable `CommandLine.sh` bash script that performs the following steps:

1. Creates a `merged_courses.tsv` adding the headers of each column and appending all the information from the 6000 `course_i.tsv` files we created in **Section 1**.

2. Takes the `country` column of the `merged_courses.tsv`, counts each time a unique country appears and prints the country that appears the **greatest** number of times. It does the same also for the `city` column.

3. Takes the `universityName` and `modality` columns of the `merged_courses.tsv`,  selects all the courses that include the word "Part time" in the `modality` column using a RegEx (`Part time`), counts each time a unique university has Part Time courses and print the **total** number of universities that offer Part Time.

4. Takes the `courseName` column of the `merged_courses.tsv`, selects all the courses that include the word Engineer/engineer using a RegEx (`[eE]ngineer`) and count the **total** number of universities.

For more information about the **implementation** of this script please refer to the `CommandLine.sh` script file included in our repository. We can test this script using the bash commands given by Jupyter and obtaining the following results:



In [9]:
#Here we run the CommandLine.sh script in our terminal
! bash CommandLine.sh


The country that offers most courses is:  United Kingdom
The city that offers most courses is:  London
The number of universities that offer part-time courses is: 121
The percentage of courses in Engineering is: 10.4167%


**NOTE:** We didn't include a screenshot of the script running in our terminal since essentially we're doing the same by running it right here in the notebook.

----

## **7. Algorithmic Question**

We were presented with the following question:

>Leonardo is an intern at a company. He is paid based on the total number of hours he has worked. They agreed __d__ days ago that Leonardo could not work less than $minTime_i$ or more than $maxTime_i$ hours per <ins>i-th</ins> day. Furthermore, he was warned by HR that on his last day at the company, he should provide a detailed report on how many hours he worked <ins>each day</ins> for the previous d days.
>
>Today is the day Leonardo should report to HR, but the problem is that he <ins>didn't</ins> account for how many hours he put in for each day, so he only has the __total sum of the hours__ ($sumHours$) he put in total in these d days. He believes that if he creates a report in which each number $dayHours_i$ corresponds to the __total hours he worked on the i-th day__ while satisfying the HR limitations and the total sum of all $dayHours_i$ equals $sumHours$, he would be fine.
>
>He cannot create such a report independently and requests your assistance. He will give you the number of days $d$, total hours spent $sumHours$, and the HR limitations for each day $i$, and he wants you to assist him in determining whether it is possible to create such a fake report. If that is possible, make such a report. 
>
>**Input**
>
>The first line of input contains two integers __d__, $sumHours$ - the number of days Leonardo worked there and the total number of hours he worked for the company. Each of the following __d__ lines contains two integer numbers $minTime_i$ and $maxTime_i$ - the minimum and maximum hours he can work on the $i_{th}$ day. 
>
>**Output**
>
>If such a report cannot be generated, print 'NO' in one output line. If such a report is possible, print 'YES' in the output and d numbers - the number of hours Leonardo spent each day - in the second line. If more than one solution exists, print any of them. 

### **7.1 Implementing a Solution**

As a first approach to this problem, we wrote a script that performs the following steps:

1. Reads the first line of input: $d$, the number of days Leonardo worked, and $sumHours$, the total number of hours Leonardo worked during the $d$ days.

2. Reads the other $d$ lines of input: $minTime$, the minimum hours Leonardo can work during the $i$-th day and $maxTime$, the maximum hours Leonardo can work during the $i$-th day.

3. Prints `NO` when one of the following conditions are satisfied:

    - The sum of the the minimum hours Leonardo can work ($sum\_ min\_ hours$) for all the $d$ days is greater than the actual hours Leonardo worked ($sumHours$).
    
    - The sum of the maximum hours Leonardo can work ($sum\_ max\_ hours$) for all the $d$ days is less than the actual hours Leonardo worked ($sumHours$).

    This is done since by definition $sum\_ min\_ hours \leq sumHours\leq sum\_ max\_ hours$.

4. Prints `YES` when $sum\_ min\_ hours \leq sumHours\leq sum\_ max\_ hours$ and finds a solution that satisfies this condition performing the following steps:

    - Obtaining the minimum hours Leonardo **should** work each day.

    - If the sum of these hours is **less** than the total hours Leonardo worked ($sumHours$) then:

        * We sum **1 hour** to one of the days (if we can) and go on to the next day.
        
        * We check if the sum of the minimum hours (plus 1) is still less than $sumHours$, if it is, we repeat the loop.

    This loop finishes when $sum\_ min\_ hours = sumHours\leq sum\_ max\_ hours$. When it does, we print the hours by day Leonardo "worked" for the condition to be satisfied.

We can see the script in the next cell, tested with the following input:

__Input 1__
```
2 5
0 1
3 5
```

In [19]:
#First, we read the first line of input, we also split it by space and we convert each element to an integer:
#d: the number of days Leonardo worked
#sumHours: the total number of hours Leonardo worked during the d days
d, sumHours = map(int, input().strip().split())

#Then, we read the other d lines of input, which we also split by space, convert each element to an integer and we store them in a list called daily_hours
#Each element of the list is a tuple (minTime, maxTime) which represents the minimum and maximum number of hours Leonardo worked during the i-th day
daily_hours = [tuple(map(int, input().strip().split())) for i in range(d)]

#Now, we calculate the sum of the minimum hours and the sum of the maximum hours that Leonardo can work during the d days
sum_min_hours = sum([minTime for minTime, maxTime in daily_hours])
sum_max_hours = sum([maxTime for minTime, maxTime in daily_hours])

#We use this to check two conditions:
#1. If the total number of hours Leonardo worked during the d days is less than the sum of the minimum hours he should work during the d days, then we print "NO" and we exit the program
#2. If the total number of hours Leonardo worked during the d days is greater than the sum of the maximum hours he should work during the d days, then we print "NO" and we exit the program
if sum_min_hours > sumHours or sum_max_hours < sumHours:
    print("NO")

#If instead, the total number of hours Leonardo worked during the d days is between the sum of the minimum hours and the sum of the maximum hours he should work during the d days, we can find a solution.
#To find the solution we perform the following steps: 
else:
    #First we print "YES" to indicate that we can find a solution
    print("YES")
    #First we initialize three variables:
    #hours: a list of length d which contains the hours Leonardo will write in the report for each day. We initialize this list with the minimum hours Leonardo has to work during each day
    #sum_hours: the sum of the hours Leonardo will write in the report for each day. We initialize this variable with the sum of the minimum hours Leonardo has to work during each day
    #i: an integer which represents the index of the day we are currently working on
    hours = [minTime for minTime, maxTime in daily_hours]
    sum_hours = sum_min_hours
    i = 0
    #Now, we start a while loop to find a solution that satisfies the condition that sum_hours == sumHours. To do this, we perform the following steps:
    while sum_hours < sumHours:
        #First we check if the minimum hours Leonardo "worked" during the i-th day is less than the maximum hours he can work during the i-th day. 
        #If this is true, then we can increase the number of hours he "worked" during the i-th day by 1 and check again if the condition sum_hours == sumHours is satisfied.
        if hours[i] < daily_hours[i][1]:
            #Here we increase the number of hours Leonardo "worked" during the i-th day by 1
            hours[i] += 1
            #Here we increase the sum of the hours Leonardo "worked" during the d days by 1
            sum_hours += 1
        #Here we increase the index of the day we are currently working on by 1 and use the modulo operator to make sure that the index is between 0 and d-1 always
        #This ensures that we are always working on a valid day
        i = (i + 1) % d
    #The while loop ends when the condition sum_hours == sumHours is satisfied. Now, we can print the solution that satisfies this condition
    #To do this, we print the elements of the hours list separated by a space
    print(*hours, sep=" ")



YES
1 4


### **7.2 Calculated Time Complexity O**

Let's calculate the time complexity (Big O notation) of this script by line:

1. The first line has **constant** time complexity $O(c)$, for $c\in \mathbf{N}$, since it only reads two values and converts them to integer:

    ```python
    d, sumHours = map(int, input().strip().split())
    ```

2. The second line has time complexity $O(d)$, where $d$ is the number of days Leonardo worked since it saves $d$ pairs of integer values in a list:

    ```python
    daily_hours = [tuple(map(int, input().strip().split())) for i in range(d)]
    ```

3. The third line has **overall** time complexity $O(d)$ since first it obtains the `minTime` values of the `daily_hours` list (with time complexity $O(d)$) and then sums these values (with time complexity $O(d)$):

    ```python
    sum_min_hours = sum([minTime for minTime, maxTime in daily_hours])
    ```

    The same thing applies for line four:

    ```python
    sum_max_hours = sum([maxTime for minTime, maxTime in daily_hours])
    ```

4. The fifth line checks if a condition is satisfied. From this line onwards, we have two options:

    - If the last condition is satisfied, then we print `NO` and exit the program. This has time complexity $O(c)$, for $c\in \mathbf{N}$:

        ```python
        if sum_min_hours > sumHours or sum_max_hours < sumHours:
            print("NO")
        ```

        When this happens, the **overall** time complexity of all the program is: $O(d)$ since it **dominates** through all the constant time complexities.

    - If it isn't satisfied, then we print `YES` (with t.c. $O(1)$) and perform extra steps since we have to find a solution for the hours Leonardo should work:

        * First we initialize a list with the minimum hours Leonardo should work for each day, this has time complexity $O(d)$ since we have to loop for all days:

            ```python
            hours = [minTime for minTime, maxTime in daily_hours]
            ```

        * We also initialize a variable that contains the sum of the hours Leonardo "worked" and an index $i$. This clearly has time complexity $O(c)$, for $c\in \mathbf{N}$:

            ```python
            sum_hours = sum_min_hours
            i = 0
            ```

        * Then, we create a `while` loop that sums 1 hour to each element of the `hours` list **until** the `sum_hours` and `sumHours` variables are the same. Let's analyze only the **worst** case scenario of this loop:

            ```python
            while sum_hours < sumHours:
                if hours[i] < daily_hours[i][1]:
                    hours[i] += 1
                    sum_hours += 1
                i = (i + 1) % d
            ```

            We can observe that this loop not only loops throughout the $d$ elements of the `hours` list, but it also can repeat this loop a particular number of times depending on the initialization of our `sum_hours` variable since we can not always enter the `if` condition. Moreover, the **worst** case scenario would be when `sum_hours = 0` since in this case we would have to loop for a minimum of `sumHours` steps (since we sum by 1 at a time) to a maximum of `sumHours`*`d` steps since we can skip the `if` a maximum of `d` number of times.

            Therefore in the **worst** case scenario the **overall** time complexity of this chunk of the program is: $O(d*sumHours)$.

        * Finally, we print the hours Leonardo "worked" each day to satisfy the `sumHours`. This takes overall time complexity $O(d)$ since we print $d$ elements:

            ```python
             print(*hours, sep=" ")
            ```


When summing the time complexities of the entire program, we can see that, in the **worst** case scenario (when we have to find a solution and sum a big number of times), our program has an **overall** time complexity of $\boxed{O(d*sumHours)}$ since it **dominates** through all the constant time complexities and $O(d)$ time complexities.


### **7.3 ChatGPT's answer**

We can ask ChatGPT to obtain our code's time complexity to check if indeed it obtains the same or not. The prompt we wrote was the following:

>Obtain the time complexity (big O notation) of the following code:
>```python
>d, sumHours = map(int, input().strip().split())
>daily_hours = [tuple(map(int, input().strip().split())) for i in range(d)]
>sum_min_hours = sum([minTime for minTime, maxTime in daily_hours])
>sum_max_hours = sum([maxTime for minTime, maxTime in daily_hours])
>
>if sum_min_hours > sumHours or sum_max_hours < sumHours:
>    print("NO")
>else:
>    print("YES")
>    hours = [minTime for minTime, maxTime in daily_hours]
>    sum_hours = sum_min_hours
>    i = 0
>    while sum_hours < sumHours:
>        if hours[i] < daily_hours[i][1]:
>            hours[i] += 1
>            sum_hours += 1
>        i = (i + 1) % d
>    print(" ".join(map(str, hours)))
>```

And its answer was:

>Let's analyze the time complexity of the given code:
>
>1. The first line takes constant time, $O(1)$, to read and parse two integers.
>
>2. The second line creates a list of tuples, which takes $O(d)$ time, where $d$ is the number of days.
>
>3. The third and fourth lines calculate the sum of the minimum and maximum hours, respectively. Both of these operations take $O(d)$ time.
>
>4. The if-else block has constant time complexity as it involves simple arithmetic operations and conditional checks.
>
>Now, let's analyze the time complexity of the while loop:
>
>- The while loop iterates until `sum_hours` is equal to or greater than `sumHours`.
>
>- In each iteration, the loop increments the `hours` array and updates the `sum_hours`.
>
>- The loop goes through each day, updating the `hours` array and `sum_hours` accordingly.
>
>The worst-case scenario is when the initial `sum_hours` is less than `sumHours`, and each day needs to be incremented until `sum_hours` becomes equal to or greater than `sumHours`. In the worst case, each day will be visited at most `sumHours` times.
>
>Therefore, the overall time complexity of the while loop is $O(d * sumHours)$.
>
>The dominant term in the time complexity is $O(d * sumHours)$, so the final time complexity of the entire code is $O(d * sumHours)$.

Surprisingly this is accurate even though it is not as well explained as we would hope so. We believe this is correct since it is the same result we arrived to by justifying every step of the way.

### **7.4 Is our code optimal?**

We believe our code is **not** optimal and can be reduced to $O(d)$ for two reasons:

1. In the **best** case scenario our code (when the first condition is satisfied) has a time complexity of $O(d)$ since it only reads the time values for each day and performs a comparison. This makes us think that we could achieve the same time complexity also when the other condition is satisfied. Indeed we can have time complexity of $O(d)$ for the second condition in the specific case when `sum_hours` is equal to `sumHours` from the beginning.

2. In the `while` loop we can not always enter the `if` condition when `hours[i] >= daily_hours[i][1]` (i.e. when the hours worked in a day are equal or greater than the maximum ones Leonardo can work). This is the reason why we can **repeat** the loop more than $d$ times and results in a waste of time since when this condition is not fulfilled, the code's running time increases a lot.

    Ideally we would like to loop through the `daily_hours` list the **minimum** number of times possible (that would be indeed $d$).

For these reasons we propose the following improvement to our previous code (tested with the same code as before):

In [20]:
#First, we read the first line of input, we also split it by space and we convert each element to an integer:
#d: the number of days Leonardo worked
#sumHours: the total number of hours Leonardo worked during the d days
d, sumHours = map(int, input().strip().split())

#Then, we read the other d lines of input, which we also split by space, convert each element to an integer and we store them in a list called daily_hours
#Each element of the list is a tuple (minTime, maxTime) which represents the minimum and maximum number of hours Leonardo worked during the i-th day
daily_hours = [tuple(map(int, input().strip().split())) for i in range(d)]

#Now, we calculate the sum of the minimum hours and the sum of the maximum hours that Leonardo can work during the d days
sum_min_hours = sum([minTime for minTime, maxTime in daily_hours])
sum_max_hours = sum([maxTime for minTime, maxTime in daily_hours])

#We use this to check two conditions:
#1. If the total number of hours Leonardo worked during the d days is less than the sum of the minimum hours he should work during the d days, then we print "NO" and we exit the program
#2. If the total number of hours Leonardo worked during the d days is greater than the sum of the maximum hours he should work during the d days, then we print "NO" and we exit the program
if sum_min_hours > sumHours or sum_max_hours < sumHours:
    print("NO")

#If instead, the total number of hours Leonardo worked during the d days is between the sum of the minimum hours and the sum of the maximum hours he should work during the d days, we can find a solution.
#To find the solution we perform the following steps: 
else:
    #First, we find the difference between the total number of hours Leonardo worked during the d days and the sum of the minimum hours he should work during the d days
    #This difference represents the number of hours we can add to the already minimum hours we know Leonardo has to work during the d days
    diff = sumHours - sum_min_hours
    #We also initialize a list called hours which will contain the hours Leonardo will write in the report for each day
    hours = []
    #Now, we iterate over the d days and perform the following steps:
    for i in range(d):
        #Here we get the minimum hours Leonardo has to work during the i-th day
        minTime = daily_hours[i][0]
        #Here we get the maximum hours Leonardo can work during the i-th day
        maxTime = daily_hours[i][1]
        #Here we get the number of hours that we can add to the i-th day. We try to optimize this number by making sure that we don't add more hours than the maximum hours Leonardo can work during the i-th day
        #We also make sure that we don't add more hours than the difference between the total number of hours Leonardo worked during the d days and the sum of the minimum hours he should work during the d days
        #We do this by taking the minimum between the difference and the maximum hours Leonardo can work during the i-th day
        add = min(diff, maxTime - minTime)
        #Here we add the number of hours we can add to the i-th day and add them to the minimum hours Leonardo has to work during the i-th day
        hours.append(minTime + add)
        #Finally we subtract the number of hours we added to the i-th day from the difference since we already used them
        diff -= add

    #In the end we should have used all the hours we can add to the already minimum hours we know Leonardo has to work during the d days
    #Nevertheless, we check if this is true by checking if the difference is greater than 0. If this is true, then we print "NO" and we exit the program
    if diff > 0:
        print("NO")
    #If instead, the difference is equal to 0, then we can print the solution that satisfies the condition diff == 0
    else:
        print("YES")
        print(*hours)
    



YES
1 4


As we can see, we only changed the second condition, since the `while` loop was the part that was more expensive in time. Instead we propose the following solution:

1. First, we find the number of hours we **have** to add to the minimum hours Leonardo has to work and also create a list to store the final result. This has constant time complexity $O(2)$:
    ```python
    diff = sumHours - sum_min_hours
    hours = []
    ```

2. Then, we start a `for` loop where we visit each day **only once**:
    ```python
    for i in range(d):
        minTime = daily_hours[i][0]
        maxTime = daily_hours[i][1]
        add = min(diff, maxTime - minTime)
        hours.append(minTime + add)
        diff -= add
    ```

    As we can see, inside the `for` loop we're only dealing with numbers (defining variables, calculating the minimum of two numbers, appending a value and substracting a value). Therefore we repeat each step only once for each day. This ensures us that the overall time complexity of this `for` loop is $O(d)$.

3. Finally, we print the result. In the worst case scenario we have to print the solution we've find, so since we have to print the hours worked for each day, this has time complexity of $O(d)$.

Therefore in this case, we have an overall time complexity for all the program of $\boxed{O(d)}$. Clearly this is **optimal** since as a minimum we have to read the values for all the $d$ days and this has the same time complexity.